File: algorithm/740. Delete and Earn/740.cpp
class Solution {
public:
    int deleteAndEarn(vector<int>& nums) {
        int n = 10001;
        vector<int> values(n, 0);
        for (int num : nums)
            values[num] += num;
        int skip = 0, take = 0;
        for(int i = 0; i < n; i++) {
            int skipi = max(take, skip);
            int takei = skip + values[i];
            skip = skipi;
            take = takei;
        }
        return max(take, skip);
    }
};

File: algorithm/740. Delete and Earn/740.java
class Solution {
    public int deleteAndEarn(int[] nums) {
        int n = 10001;
        int[] values = new int[n];
        for(int i = 0; i < nums.length; i++) {
            values[nums[i]] += nums[i];
        }
        
        int take =0, skip = 0;
        for(int i = 0; i < n; i++) {
            int takei = skip + values[i];
            int skipi = Math.max(skip, take);
            skip = skipi;
            take = takei;
        }
        return Math.max(skip, take);
    }
}
